integration-tests/tests/snap_op_test.go

```diff
@@ -76,3 +80,52 @@ func (s *snapOpSuite) TestRemoveRemovesAllRevisions(c *check.C) {
 	c.Assert(err, check.IsNil)
 	c.Check(revnos, check.HasLen, 0)
 }
+
+// TestRemoveBusyRetries is a regression test for LP:#1571721
+func (s *snapOpSuite) TestRemoveBusyRetries(c *check.C) {
+	// install the binaries snap
+	snapPath, err := build.LocalSnap(c, data.BasicBinariesSnapName)
+	defer os.Remove(snapPath)
+	c.Assert(err, check.IsNil, check.Commentf("Error building local snap: %s", err))
+	installOutput := installSnap(c, snapPath)
+	c.Assert(installOutput, testutil.Contains, data.BasicBinariesSnapName)
+
+	// run a command that keeps the mount point busy in the background
+	blockerBin := fmt.Sprintf("/snap/bin/%s.block", data.BasicBinariesSnapName)
+	blockerSrv := "umount-blocker"
+	cli.ExecCommand(c, "sudo", "systemd-run", "--unit", blockerSrv, blockerBin)
+	c.Assert(err, check.IsNil, check.Commentf("cannot start %q:%s", blockerBin, err))
+	wait.ForActiveService(c, blockerSrv)
+
+	// snap remove will block (and try to retry) while the umount-block
+	// service is running. so we need to do stuff in a go-routine
+	ch := make(chan int)
+	go func() {
+		// wait until snappy to show that it is retrying
+		needle := `Doing.*Remove`
+		wait.ForCommand(c, needle, "snap", "changes")
+
+		// find change id of the remove
+		output := cli.ExecCommand(c, "snap", "changes")
+		id := regexp.MustCompile(`(?m)([0-9])+.*Doing.*Remove.*"`).FindStringSubmatch(output)[1]
+		needle = `retrying`
+		wait.ForCommand(c, needle, "snap", "changes", id)
+
+		// now stop the service that blocks the umount
+		cli.ExecCommand(c, "sudo", "systemctl", "stop", blockerSrv)
+		wait.ForInactiveService(c, blockerSrv)
+
+		// this triggers an Ensure in the overlord
+		cli.ExecCommandErr("sudo", "snap", "refresh", "ubuntu-core")
+		ch <- 1
+	}()
+
+	// try to remove, this will block and eventually succeed
+	_, err = cli.ExecCommandErr("sudo", "snap", "remove", data.BasicBinariesSnapName)
+	c.Check(err, check.IsNil)
+	<-ch
+
+	// ensure no changes are left in Doing state
```

Comment:
Shouldn't that be above the channel receive? Otherwise the test will probably pass as a side effect of the channel receive, rather than snap remove blocking until done.